src/aiq/front_ends/mcp/mcp_front_end_plugin.py

```diff
@@ -0,0 +1,98 @@
+from aiq.runtime.session import AIQSessionManager
+from typing import Dict
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class MCPFrontEndPlugin(FrontEndBase[MCPFrontEndConfig]):
+    """MCP front end plugin implementation."""
+
+    async def run(self) -> None:
+        """Run the MCP server."""
+        # Import FastMCP
+        from mcp.server.fastmcp import FastMCP, Context
+
+        # Create an MCP server with the configured parameters
+        mcp = FastMCP(
+            self.front_end_config.name,
+            host=self.front_end_config.host,
+            port=self.front_end_config.port,
+            debug=self.front_end_config.debug,
+            log_level=self.front_end_config.log_level,
+        )
+
+        # Build the workflow and register all functions with MCP
+        async with WorkflowBuilder.from_config(config=self.full_config) as builder:
+            # Build the workflow
+            workflow = builder.build()
+
+            # Get all functions from the workflow
+            functions = self._get_all_functions(workflow)
+
+            # Filter functions based on tool_names if provided
+            if self.front_end_config.tool_names:
+                logger.info(f"Filtering functions based on tool_names: {self.front_end_config.tool_names}")
+                filtered_functions: Dict[str, Function] = {}
+                for function_name, function in functions.items():
+                    if function_name in self.front_end_config.tool_names:
+                        filtered_functions[function_name] = function
+                    else:
+                        logger.debug(f"Skipping function {function_name} as it's not in tool_names")
+                functions = filtered_functions
+
+            # Register each function with MCP
+            for function_name, function in functions.items():
+                register_function_with_mcp(mcp, function_name, function)
+
+            # Add a simple fallback function if no functions were found
+            if not functions:
+                logger.warning("No functions found in workflow, adding placeholder function")
+
+                @mcp.tool()
+                def add(a: int, b: int) -> int:
+                    """Add two numbers."""
+                    return a + b
+
+            # Start the MCP server
+            await mcp.run_sse_async()
+
+    def _get_all_functions(self, workflow: Workflow) -> Dict[str, Function]:
+        """Get all functions from the workflow.
+
+        Args:
+            workflow: The AIQ workflow.
+
+        Returns:
+            Dict mapping function names to Function objects.
+        """
+        functions: Dict[str, Function] = {}
+
+        # Extract all functions from the workflow
+        for function_name, function in workflow.functions.items():
+            functions[function_name] = function
+
+        functions[workflow.config.workflow.type] = workflow
```

Comment:
The function name for the workflow should be set to the package name (type can be the same for many workflows). This can be done in a separate PR,

Reply:
Correct 2 is better but it is not needed to merge this MR.